File: apps/frontend/src/lib/api/share-links.ts
/**
 * Agent Share Links API
 *
 * Handles public sharing of agents via unique share tokens.
 */

import { backendApi } from '../api-client';
import { handleApiError } from '../error-handler';

// ============================================================================
// Types
// ============================================================================

export interface ShareLinkSettings {
  rate_limit_per_hour?: number;
  allow_file_access?: boolean;
  custom_greeting?: string;
}

export interface ShareLinkCreateRequest {
  expires_in_days?: number;
  settings?: ShareLinkSettings;
}

export interface ShareLinkAgentInfo {
  agent_id: string;
  name: string;
  description?: string;
  icon_name?: string;
  icon_color?: string;
  icon_background?: string;
}

export interface ShareLink {
  share_id: string;
  agent_id: string;
  created_at: string;
  expires_at?: string;
  is_active: boolean;
  views_count: number;
  runs_count: number;
  last_viewed_at?: string;
  last_run_at?: string;
  settings?: ShareLinkSettings;
}

export interface ShareLinksListResponse {
  share_links: ShareLink[];
}

export interface PublicShareLink {
  share_id: string;
  agent: ShareLinkAgentInfo;
  views_count: number;
  settings?: ShareLinkSettings;
}

export interface ShareLinkError {
  error: string;
  code: 'LINK_DEACTIVATED' | 'LINK_EXPIRED' | 'LINK_NOT_FOUND' | 'AGENT_NOT_FOUND';
}

export interface ShareLinkUpdateRequest {
  is_active?: boolean;
  settings?: ShareLinkSettings;
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Create a new share link for an agent
 */
export async function createShareLink(
  agentId: string,
  request: ShareLinkCreateRequest = {}
): Promise<ShareLink> {
  try {
    const response = await backendApi.post(`/agents/${agentId}/share-links`, request);
    return response.data;
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * List all share links for an agent
 */
export async function listShareLinks(agentId: string): Promise<ShareLinksListResponse> {
  try {
    const response = await backendApi.get(`/agents/${agentId}/share-links`);
    return response.data;
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * Get public share link details (no auth required)
 */
export async function getPublicShareLink(shareId: string): Promise<PublicShareLink> {
  try {
    const response = await backendApi.get(`/share/${shareId}`);
    return response.data;
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * Update a share link
 */
export async function updateShareLink(
  shareId: string,
  request: ShareLinkUpdateRequest
): Promise<ShareLink> {
  try {
    const response = await backendApi.patch(`/share-links/${shareId}`, request);
    return response.data;
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * Delete a share link permanently
 */
export async function deleteShareLink(shareId: string): Promise<void> {
  try {
    await backendApi.delete(`/share-links/${shareId}`);
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * Revoke (deactivate) a share link
 */
export async function revokeShareLink(shareId: string): Promise<ShareLink> {
  try {
    const response = await backendApi.post(`/share-links/${shareId}/revoke`);
    return response.data;
  } catch (error) {
    throw handleApiError(error);
  }
}

/**
 * Generate the public URL for a share link
 */
export function getShareLinkUrl(shareId: string): string {
  // Get the base URL from the current window location
  const baseUrl = typeof window !== 'undefined'
    ? `${window.location.protocol}//${window.location.host}`
    : '';
  return `${baseUrl}/share/agent/${shareId}`;
}

File: backend/core/api_models/share_links.py
"""Agent share link API models."""

from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime


class ShareLinkSettings(BaseModel):
    """Settings for a share link."""
    rate_limit_per_hour: Optional[int] = 10  # Max runs per hour for public users
    allow_file_access: Optional[bool] = False  # Allow file upload/download
    custom_greeting: Optional[str] = None  # Custom welcome message


class ShareLinkCreateRequest(BaseModel):
    """Request model for creating a share link."""
    expires_in_days: Optional[int] = None  # Optional expiration in days
    settings: Optional[ShareLinkSettings] = None


class ShareLinkAgentInfo(BaseModel):
    """Basic agent info for share link responses."""
    agent_id: str
    name: str
    description: Optional[str] = None
    icon_name: Optional[str] = None
    icon_color: Optional[str] = None
    icon_background: Optional[str] = None


class ShareLinkResponse(BaseModel):
    """Response model for a share link."""
    share_id: str
    agent_id: str
    created_at: str
    expires_at: Optional[str] = None
    is_active: bool
    views_count: int
    runs_count: int
    last_viewed_at: Optional[str] = None
    last_run_at: Optional[str] = None
    settings: Optional[Dict[str, Any]] = None


class ShareLinksListResponse(BaseModel):
    """Response model for list of share links."""
    share_links: List[ShareLinkResponse]


class PublicShareLinkResponse(BaseModel):
    """Response model for public share link access."""
    share_id: str
    agent: ShareLinkAgentInfo
    views_count: int
    settings: Optional[Dict[str, Any]] = None


class PublicShareLinkErrorResponse(BaseModel):
    """Response model for share link errors."""
    error: str
    code: str  # LINK_DEACTIVATED, LINK_EXPIRED, AGENT_NOT_FOUND


class ShareLinkUpdateRequest(BaseModel):
    """Request model for updating a share link."""
    is_active: Optional[bool] = None
    settings: Optional[ShareLinkSettings] = None
